ambry-router/src/test/java/com/github/ambry/router/PutOperationTest.java

```diff
@@ -969,4 +974,62 @@ public void compressChunk()
     ByteBuf buf = (ByteBuf) FieldUtils.readField(putChunk, "buf", true);
     buf.release();
   }
+
+  /**
+   * Ensure that if any of the requests associated with the buffer of a PutChunk is not completely read out even
+   * after the associated chunk is complete, the buffer is not reused even though the PutChunk is reused.
+   */
+  @Test
+  public void testRetry() throws Exception {
+    int numChunks = routerConfig.routerMaxInMemPutChunks + 1;
+    BlobProperties blobProperties =
+        new BlobProperties(-1, "serviceId", "memberId", "contentType", false, Utils.Infinite_Time,
+            Utils.getRandomShort(TestUtils.RANDOM), Utils.getRandomShort(TestUtils.RANDOM), false, null, null, null);
+    byte[] userMetadata = new byte[10];
+    byte[] content = new byte[chunkSize * numChunks];
+    random.nextBytes(content);
+    ReadableStreamChannel channel = new ByteBufferReadableStreamChannel(ByteBuffer.wrap(content));
+    FutureResult<String> future = new FutureResult<>();
+    MockNetworkClient mockNetworkClient = new MockNetworkClient();
+    MockClusterMap mockClusterMap = spy(new MockClusterMap());
+    MockPartitionId mockPartitionIdFail = new MockPartitionId();
+    MockPartitionId mockPartitionIdPass =
+        (MockPartitionId) mockClusterMap.getAllPartitionIds(MockClusterMap.DEFAULT_PARTITION_CLASS).get(1);
+
+    AtomicInteger callCount = new AtomicInteger(0);
+    // Mock the behavior of getRandomWritablePartition to return different values on each call
+    doAnswer(invocation -> {
+      int callNum = callCount.getAndIncrement();  // Increment call count and get the current value
+      if (callNum == 0) {
+        return mockPartitionIdFail;  // First call returns the first partition
+      } else {
+        return mockPartitionIdPass;  // Second call returns the second partition
+      }
+    }).when(mockClusterMap).getRandomWritablePartition(any(), any());
```

Comment:
we should do two tests, first test just returns mockPartitionIdFail in both cases, and we expect to see exception for op, second test returns different ids, and we expect to see success for op.